### PR TITLE
feat: support load-update-data

### PR DESCRIPTION
**Pull request by olavloite**
Adds support for the load-update-data change type. This change type normally uses an UPSERT / MERGE statement. Cloud Spanner does not support that for DML statements, and Liquibase does not support using anything else than SQL for this change type. That means that the change type for Cloud Spanner will always generate two statements for each row that will be inserted/updated:
1. An optional `INSERT` statement that will check whether the row already exists.
1. An `UPDATE` statement that always updates the row (even if it was just inserted).

The more efficient way to do this would be to use an `InsertOrUpdate` mutation in Cloud Spanner. There is however currently no way to trigger that through SQL.

### CODE REVIEW

File: src/main/java/liquibase/ext/spanner/SpannerInsertOrUpdateGenerator.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+  @Override
+  protected String getInsertStatement(InsertOrUpdateStatement insertOrUpdateStatement,
+      Database database, SqlGeneratorChain sqlGeneratorChain) {
+    InsertWithSelectGenerator insertGenerator = new InsertWithSelectGenerator();
+    StringBuffer sql = new StringBuffer(
+        insertGenerator.generateSql(insertOrUpdateStatement, database, sqlGeneratorChain)[0]
+            .toSql());
+    sql
+        .append(" FROM UNNEST([1])") // only SELECT statements with a FROM may have a WHERE clause.
```
Comment (mescanne) [inline src/main/java/liquibase/ext/spanner/SpannerInsertOrUpdateGenerator.java:75]: This is funny. So you must have a FROM clause because a WHERE clause would normally only be applied to the source table. However, in this case the WHERE clause is applied to a completely different table than you're selecting from.

File: src/main/java/liquibase/ext/spanner/SpannerInsertOrUpdateGenerator.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      sqlList.add(
+          new UnparsedSql(getInsertStatement(insertOrUpdateStatement, database, sqlGeneratorChain),
+              "", getAffectedTable(insertOrUpdateStatement)));
+    }
+
+    String whereClause = getWhereClause(insertOrUpdateStatement, database);
+    try {
+      String update =
+          getUpdateStatement(insertOrUpdateStatement, database, whereClause, sqlGeneratorChain);
+      if (update.endsWith("\n") && update.length() > 1) {
```
Comment (mescanne) [inline src/main/java/liquibase/ext/spanner/SpannerInsertOrUpdateGenerator.java:55]: Some of this code I assume is taken from similar functions in core Liquibase? This is yet another way to remove a character from an end of a string.

File: src/main/java/liquibase/ext/spanner/InsertWithSelectGenerator.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        }
+      } else if (newValue instanceof DatabaseFunction) {
+        sql.append(database.generateDatabaseFunctionValue((DatabaseFunction) newValue));
+      } else {
+        sql.append(newValue);
+      }
+      sql.append(", ");
+    }
+
+    sql.deleteCharAt(sql.lastIndexOf(" "));
```
Comment (mescanne) [inline src/main/java/liquibase/ext/spanner/InsertWithSelectGenerator.java:82]: So we rely on the last character always being a space, but need to check to see if it's a comma -- it doesn't make much sense to me. I assume this is taken from other Liquibase code.

**Review by mescanne — APPROVED** [review]
Some oddities in there but the tests are thorough. I can see potential issues coming through edge cases in the future.

File: src/main/java/liquibase/ext/spanner/SpannerInsertOrUpdateGenerator.java (hunk 2 above)
Comment (olavloite) [reply src/main/java/liquibase/ext/spanner/SpannerInsertOrUpdateGenerator.java:55]: Yes. I tried to reuse the original `InsertOrUpdateGenerator`, but the (internal) API of it changed a little bit between version 3.x and 4.x, so it's not doable if we want to be able to support both. So I ended up copying the main parts of the SQL generation into this generator.

File: src/main/java/liquibase/ext/spanner/InsertWithSelectGenerator.java (hunk 3 above)
Comment (olavloite) [reply src/main/java/liquibase/ext/spanner/InsertWithSelectGenerator.java]: Yeah, this one is too strange. I've sanitized this generator a little bit (but yes, it was mostly re-use of code from the original generator).
